linkup-cli/src/completion.rs

```diff
@@ -0,0 +1,20 @@
+use std::io::stdout;
+
+use clap::{Command, CommandFactory};
+use clap_complete::{generate, Generator, Shell};
+
+use crate::{Cli, CliError};
+
+pub fn completion(shell: &Option<Shell>) -> Result<(), CliError> {
+    if let Some(shell) = shell.clone() {
+        let mut cmd = Cli::command();
+        print_completions(shell, &mut cmd);
+        Ok(())
+    } else {
+        Ok(())
+    }
+}
+
+fn print_completions<G: Generator>(gen: G, cmd: &mut Command) {
+    generate(gen, cmd, cmd.get_name().to_string(), &mut stdout());
+}
```

Comment:
**_suggestion/nit:_**
Cool stuff! We can probably move the clone responsibility to the printing method. Something like this:
```suggestion
pub fn completion(shell: &Option<Shell>) -> Result<(), CliError> {
    if let Some(shell) = shell {
        let mut cmd = Cli::command();
        print_completions(shell, &mut cmd);
        Ok(())
    } else {
        Ok(())
    }
}

fn print_completions<G: Generator + Clone>(gen: &G, cmd: &mut Command) {
    generate(gen.clone(), cmd, cmd.get_name().to_string(), &mut stdout());
}
```

Reply:
Nice!

As an answer to both of you:
The else in the original code is essentially dead code as far as I can tell, since clap checks the shell Enum before passing stuff on to this function.

Reply:
@ostenbom fyi